ACCUMULO-1755: Modified TSBW so that all client threads will not bloc…

dlmarion opened the pull request:
I ditched review board and went back to the drawing board.

Before this change all client threads that were adding mutations to a BatchWriter would block
when the mutations needed to be written to the tablet servers. This change gives each client
thread their own MutationWriter object using a ThreadLocal. With this change only one client
should block on adding the mutations to the sendThreadPool object, and allows the other client
threads to push mutations onto a new MutationSet.

keith-turner commented:
I think this change will result in a RPC per thread using the batch writer.    If 3 threads add mutations for the same tablet server, then it will no longer send all of those in one RPC.   If 100 threads add mutations for 100 tablets server, the batch writer will make ~10,000 RPCs instead of ~100.

I have some idea on how to improve the batch writer based on what I learned while writing the conditional writer, I will try to write those up.  It may or may not be useful.

dlmarion commented:
The set of mutations is still global, if 3 threads add a mutation for the same tablet server and the MutationSet has not been swapped out (in startProcessing), then those mutations will be in the same RPC.

keith-turner commented:
> The set of mutations is still global,

OIC.  Your right, its not as bad as I thought.  Although there is still problem.  With the per thread mutation writers, mutations bound for the same tserver may never be merged in some cases.  This situations are less likely to occur though.  I am thinking of something like the following.

 * Thread 1 bins 1000 mutations for 20 tservers
 * 3 tserver queues are processed by send threads
 * Thread 2 bins 1000 mutations for 20 tserver (at this point 17 tservers from the 1st binning are not processed by send threads, those mutations are not merged together).

This makes me wonder, why have a mutation writer per thread?  Why not just add to the mutation writer outside of the sync?

dlmarion commented:
Talked with Keith offline and he suggested a simpler route. New changes uploaded.

joshelser commented on core/src/main/java/org/apache/accumulo/core/client/impl/TabletServerBatchWriter.java:
```diff
@@ -699,7 +701,16 @@ else if (Tables.getTableState(context.getInstance(), table) == TableState.OFFLIN
 
-    void addMutations(MutationSet mutationsToSend) {
+    void queueMutations(final MutationSet mutationsToSend) {
+      queueThreadPool.submit(new Runnable() {
```
I think it be simpler to use a producer/consumer model for the `MutationSet`'s with a single consumer thread instead of constantly submitting new `Runnable`'s. The consumer thread can essentially do the same thing: poll for a `MutationSet`, and then run `addMutations` with that object. It eliminates the need for another ExecutorService and repeatedly creating new Runnables. We might also be able to implement some back-pressure on the client via that Queue as well.

dlmarion replied:
Are you suggesting re-using the existing ExecutorService?

joshelser replied:
> Are you suggesting re-using the existing ExecutorService?

Nope. I'm saying just make a single thread which is reading off some queue. The clients would be submitting mutations onto that queue. You just have one thread always running, looking for mutations to add. You avoid the whole creating, scheduling and starting the Runnable.

keith-turner replied:
There is single thread.  The ExecutorService has one thread that waiting on a queue.

joshelser replied:
I didn't say there were multiple threads, there are multiple Runnables which need to be executed on that thread. This is what is unnecessary IMO

keith-turner replied:
> I didn't say there were multiple threads

Didn't intend to imply you said that.   Was comparing a thread pool with 1 thread to the following statement.

> You just have one thread always running, looking for mutations to add.

joshelser replied:
Keith was chattign with me on IRC about this. I've pseudocoded what I think would be a cleaner approach since we seem to be talking past each other:

https://paste.apache.org/0GY6

joshelser replied:
Ultimately, I really don't have that strong of an opinion here. I think using the ExecutorService as a queue for MutationSets is a smell (instead of directly maintaining such a queue), but I can understand the argument for using the ExecutorService *because* it encompasses both approaches. I don't think there's a big reason one way or the other; I simply think that the pseudo-code outline above is clearer than Dave's current changes. If Keith and Dave are happy, then merge it and move on.

dlmarion replied:
Taking your idea and going further, if we are only using one thread then we can use a timer driven approach and create zero Runnable objects. I'll update the code shortly.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/client/impl/TabletServerBatchWriter.java:
```diff
@@ -699,7 +701,16 @@ else if (Tables.getTableState(context.getInstance(), table) == TableState.OFFLIN
+    void queueMutations(final MutationSet mutationsToSend) {
+      queueThreadPool.submit(new Runnable() {
+        public void run() {
+          addMutations(mutationsToSend);
```
Need to ensure any exceptions are handled properly. If an expected exception happens here, mutations are lost, and the batch writer keeps working thats not good.

keith-turner replied:
I meant unexpected exception

keith-turner commented on core/src/main/java/org/apache/accumulo/core/client/impl/TabletServerBatchWriter.java:
```diff
@@ -629,6 +630,7 @@ public void run() {
 
     private static final int MUTATION_BATCH_SIZE = 1 << 17;
     private final ExecutorService sendThreadPool;
+    private final ExecutorService queueThreadPool = new SimpleThreadPool(1, "QueueMutations");
```
I think a name with mutations and binning or binner in it would be more useful.

dlmarion commented:
Addressed Keiths comments in the update. Waiting for resolution on Josh's comment.

keith-turner commented:
> Addressed Keiths comments in the update.

What about comment about exceptions?

dlmarion commented:
I added a catch clause to startProcessing to catch the RejectedExecutionException that can be thrown. Were you thinking of something else?

keith-turner commented:
> I added a catch clause to startProcessing to catch the RejectedExecutionException that can be thrown. Were you thinking of something else?

What you did is needed, but I was thinking of something else.   If the call to `addMutations()` in the Runnable throws an unexpected exception, nothing will be done about that exception. The batch writer should transition to an error state.  Thinking need to do something like the following.

```java
void queueMutations(final MutationSet mutationsToSend) {
      queueThreadPool.submit(new Runnable() {
        public void run() {
          try{
            addMutations(mutationsToSend);
          } catch (Exception e){
            updateUnknownErrors(e);
          }
        }
      });
    }

```

dlmarion commented:
Updated from comments by Keith and Josh.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/client/impl/TabletServerBatchWriter.java:
```diff
@@ -638,6 +648,21 @@ public MutationWriter(int numSendThreads) {
+            m = queue.poll();
+          }
+        }
+      }, 0, 500);
```
Does this mean every 500ms check for something on the queue to bin?  Whats the rational for this?

dlmarion replied:
Yes, wake up every 1/2 second and bin mutations if there are any. No rationale, just thought it made sense. You have a different approach?

keith-turner replied:
Ok I think mutations should be binned and sent immediately.   There are currently two cases when mutations are submitted to be binned and sent: when the users calls flush and when memory 1/2 full.  Don't see a good reason to wait in either case.

dlmarion replied:
Well, in the previous revision I was adding a MutationSet to an ExecutorService that had an unbounded queue and 1 thread. That's not immediate. Can MutationSets be binned concurrently? If so, then I can change the timer from a fixed delay to fixed rate. Thoughts?

keith-turner replied:
> Well, in the previous revision I was adding a MutationSet to an ExecutorService that had an unbounded queue and 1 thread. That's not immediate

Yeah, its certainly not immediate.   However, it probably much less than 1ms between adding to thread pool and thread calling run() on Runnable (when thread is idle).   Relative to ~250ms, its like kinda immediate.

> Can MutationSets be binned concurrently?

I have no idea.  Could not say yes w/o much more review of the code than I want to do late on a Fri afternoon :).  I suspect something would blow up.

dlmarion replied:
LOL, fair enough. I'm going to leave this as-is for the weekend. We can either leave it single threaded and lower the period, or change it to another Executor with multiple threads reading off a bounded/unbounded queue.

joshelser commented:
Thanks for doing this, @dlmarion. I think this approach is much easier to follow.

dlmarion commented:
Changing to unbounded is easy. Leaving at 10 (or some other number) provides back pressure to the client threads adding mutations. It's not 10 mutations, it's 10 MutationSets.

keith-turner commented:
> Leaving at 10 (or some other number) provides back pressure to the client threads adding mutations. 

I beilieve doing this makes it harder to understand the batch writer from a user perspective.   This batch writer already provides back pressure based on memory usage.  This constraint would provide back pressure in another way that I would not want to attempt to write up in user facing javadocs.

dlmarion commented:
Updated based on Keiths concerns, changed implementation such that:

1. MutationWriter uses a ThreadPoolExecutor with 1 thread instead of a Timer. We might be able to change this later, not sure.
2. The ThreadPoolExecutor uses a LinkedTransferQueue as its work queue
3. MutationWriter.queueMutations attempts to put the MutationSet onto the work queue with no timeout. If there is a consumer available, then the MutationSet will get put onto the work queue and processed by a background thread. If no consumer is available (e.g. it's currently busy), then the MutationSet will be binned in the current thread.

I also changed some variables to Atomic objects and removed the synchronization modifier from TSBW.updateBatchStats()

keith-turner commented on core/src/main/java/org/apache/accumulo/core/client/impl/TabletServerBatchWriter.java:
```diff
@@ -629,6 +639,8 @@ public void run() {
     private static final int MUTATION_BATCH_SIZE = 1 << 17;
     private final ExecutorService sendThreadPool;
+    private final LinkedTransferQueue<Runnable> queue = new LinkedTransferQueue<>();
+    private final ThreadPoolExecutor executor = new ThreadPoolExecutor(1, 1, 10L, TimeUnit.SECONDS, queue);
```
Would be nice to name thread.  Also it should be a daemon thread (I think it will not be by default).  These are things I think that SimpleThreadPool does (may have name wrong, you were using it earlier).

keith-turner commented on core/src/main/java/org/apache/accumulo/core/client/impl/TabletServerBatchWriter.java:
```diff
@@ -699,7 +711,33 @@ else if (Tables.getTableState(context.getInstance(), table) == TableState.OFFLIN
+      });
+      if (!transferred) {
+        try {
+          addMutations(mutationsToSend);
```
I like this.  Its an interesting approach of having multiple threads bin if it happens to be busy.  All batchwriter code was written w/ assumption that one thread would bin, so this change needs careful review of all code called by binning code.

dlmarion replied:
Agreed. I believe that binning can be done concurrently. There is still a synchronization point later, after the binning, but I believe that is the right place to block. I updated this PR to make the Executor a daemon with a name.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/client/impl/TabletServerBatchWriter.java:
```diff
@@ -699,7 +710,33 @@ else if (Tables.getTableState(context.getInstance(), table) == TableState.OFFLIN
+    void queueMutations(final MutationSet mutationsToSend) throws InterruptedException {
+      if (null == mutationsToSend)
+        return;
+      boolean transferred = queue.tryTransfer(new Runnable() {
```
I was playing around with this locally.  I added some prints and noticed the background thread was never binning.  This was because the background thread was never started (adding vial thread pool API will start it).  I fixed this issue by prestarting the threads, however  I think it would be safer to add using the Executor API (because threads could idle time out).  I was reading and found the following.

 * ThreadPoolExecutor javadocs recommend only using the queue for debugging and monitoring
 * The same thing can be accomplished using ThreadPoolExecutor.CallerRunsPolicy() and a SynchronousQueue()

I got .CallerRunsPolicy+SynchronousQueue working here keith-turner/accumulo@bb28195

I made some modifications to BatchWriterFlushIT in keith-turner/accumulo@3d3c252 .  Running that I was able to see that sometimes the background thread binned and sometimes the foreground thread binned.

dlmarion commented:
Realized I left a System.out.println in the test. Will remove locally and wait for feedback before pushing that change.

dlmarion commented:
Thanks josh, made the changes locally.

joshelser commented:
This is looking really nice, @dlmarion. :+1: from me.

keith-turner commented:
Since MutationWriter.getLocator()  is no longer called in sync, race conditions can occur.  Could make the entire method synchronize on locators.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/client/impl/TabletServerBatchWriter.java:
```diff
@@ -427,11 +437,11 @@ public void updateBinningStats(int count, long time, Map<String,TabletServerMuta
-    minTabletServersBatch = Math.min(minTabletServersBatch, binnedMutations.size());
-    maxTabletServersBatch = Math.max(maxTabletServersBatch, binnedMutations.size());
+    minTabletServersBatch.set(Math.min(minTabletServersBatch.get(), binnedMutations.size()));
+    maxTabletServersBatch.set(Math.max(maxTabletServersBatch.get(), binnedMutations.size()));
```
This method of updating has a race condition.   Multiple threads could call get() before calling set().  Also all of these atomic vars require round trips to main memory (not sure how much this matters).   I can think of two possible solutions.  Both involve creating a BatchWriterStats class to make the code more managable.

 1. Could add a syncrhonized updateBatchStats method to BatchWriterStats.  No longer syncing on main lock or making lots of trips to main mem.
 2. Could have an AtomicRef<BatchWriterStats>.  To update batch writer stats read the ref, clone it, make updates to clone, update ref using CAS to ensure ref has not changed.  If ref changed, then start over.  This avoids lock, race conditions, and lots of trips to main memory.

dlmarion commented:
So, I took a different approach. I believe that I resolved the race conditions by synchronizing on the objects being updated. This would still cause the performance penalty that you are talking about going to main memory. However, the stats objects being updated are only used if trace logging is enabled, so I guarded the two methods that update the stats with trace logging checks. Therefore, you will only pay a performance penalty if trace logging is enabled, but by turning on trace logging you should expect a little bit of a performance hit anyway.

dlmarion commented:
I looked into the build failure, which is a findbugs issue, and I don't see the issue that it's complaining about. AtomicInteger / Long do not implement Lock.

joshelser commented:
> which is a findbugs issue, and I don't see the issue that it's complaining about

Can you reproduce it locally via `mvn verify -DskipTests -Dcheckstyle.skip`? I know the jenkins output can sometimes be... a little weird to parse for w/e reason.

keith-turner commented:
To make findbugs happy could CAS in a loop to compute the min and max, something like :

```java
   private static void computeMin(AtomicInt stat, int update) {
       int old = stat.get();
       while(!stat.compareAndSet(old, Math.min(old, update))){
           old = stat.get();
       }
   }
```

keith-turner commented:
Java 8 added accumulateAndGet to AtomicInt which can be used w/ lambdas to compute min max.  Java 8 is so nice, but we can't use it yet.

In java 8 could do the following

```java
   atomicInt.accumulateAndGet(update, Math::min)
```

keith-turner commented:
>  I guarded the two methods that update the stats with trace logging checks. 

thats a nice improvement.

I think this patch looks good now +1

dlmarion commented:
Will apply manually. Thx.
